indra/newview/llappviewer.cpp

```diff
@@ -1375,6 +1378,51 @@ bool LLAppViewer::frame()
 	return ret;
 }
 
+
+// static
+bool packGameControlInput(LLMessageSystem* msg)
+{
+    if (!LLGameControl::hasInput())
+    {
+        return false;
+    }
+    if (!gSavedSettings.getBOOL("EnableGameControlInput"))
+    {
+        LLGameControl::clearInput();
+        return false;
+    }
+
+    msg->newMessageFast(_PREHASH_GameControlInput);
+    msg->nextBlock("AgentData");
+    msg->addUUID("AgentID", gAgentID);
+    msg->addUUID("SessionID", gAgentSessionID);
+
+    const LLGameControl::State& state = LLGameControl::getState();
+
+    size_t num_indices = state.mAxes.size();
+    for (U8 i = 0; i < num_indices; ++i)
+    {
+        if (state.mAxes[i] != state.mPrevAxes[i])
+        {
+            // only pack an axis if it differes from previously packed value
+            msg->nextBlockFast(_PREHASH_AxisData);
+            msg->addU8Fast(_PREHASH_Index, i);
+            msg->addS16Fast(_PREHASH_Value, state.mAxes[i]);
+        }
+    }
+
+    const std::vector<U8>& buttons = state.mPressedButtons;
+    if (!buttons.empty())
+    {
+        msg->nextBlockFast(_PREHASH_ButtonData);
+        msg->addBinaryDataFast(_PREHASH_Data, (void*)(buttons.data()), (S32)(buttons.size()));
```

Comment:
I think this is a bit too inflecible.
1. Not all functions are server sided, should be possible to map buttons to viewer functions, not just sent to server and let sript sort it out.
2. What if user has some kind of flight controller with minimal amount of buttons? Normally keyboard picks the slack and sends extra functions.
3. What about users that don't have controllers at all? 3p viewers will start abusing and emulating this to get access to these functions and get extended control from scripts.

I think joystick should be it's own thing and we need an extention of llTakeControls (mControlsTakenCount) that doesn't care about joysticks specifically, just allows more inputs with things like axises included.
So that instead of doing strictly 'joystick->server'. it whould be something like 'any input ->map to desired value->server'